### DISCUSSION TITLE
테스트 대역에는 어떠한 종류가 있고, 어떤 상황에서 사용하나요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#41] [Test]
.

**Comment by kunsanglee** [comment]
테스트 과정에서 테스트 더블(Test Double)은 실제 객체를 대신하여 테스트의 특정 조건을 만족시키거나 특정 상황을 시뮬레이션하는 데 사용됩니다. 왜냐하면 실제 객체를 사용하기 어렵거나 테스트 실행 시간을 단축하기 위해서입니다.

### 테스트 대역 종류와 활용 예시
테스트에서 사용되는 테스트 더블에는 주로 스텁(Stub), 모크(Mock), 스파이(Spy), 페이크(Fake) 등이 있습니다. 각각은 테스트의 특정 목적을 위해 설계되었습니다.

- 스텁(Stub)은 호출된 메서드에 대해 미리 정의된 응답을 반환하는 간단한 구현입니다. 주로 의존성이 있는 외부 시스템의 응답을 시뮬레이션할 때 사용됩니다. 예를 들면, 네트워크를 통한 데이터 요청에서 사용할 수 있습니다.

- 모크(Mock)는 호출된 메서드에 대한 기대치를 설정하고, 실행된 메서드의 호출 여부, 호출 순서, 전달된 파라미터 등을 검증하는 데 사용됩니다. 객체의 행위를 검증할 때 유용합니다.

- 스파이(Spy)는 실제 객체의 기능을 그대로 사용하면서 추가적으로 메서드 호출 정보를 기록합니다. 이 정보를 기반으로 함수가 예상대로 동작하는지 검증할 수 있습니다.

- 페이크(Fake)는 실제 동작하는 구현체이지만, 생산용에 비해 간소화되거나 더 빠르게 동작하도록 설계되었습니다. 예를 들어, 인 메모리 데이터베이스를 활용하여 데이터베이스 시스템을 시뮬레이션할 수 있습니다.

**Comment by seunghye218** [comment]
테스트 대역은 실제 의존성과 유사하게 보이고 동작하지만 실제로는 복잡성을 줄이고 테스트를 용이하게 하는 단순화된 버전인 객체를 뜻하는 용어입니다.
테스트 대역은 테스트 더블이라고도 불리며 종류로 목, 스텁, 페이크가 있습니다.

목은 테스트 대상 코드가 의존성을 통해 제공되는 함수를 호출하는 지 검증하기 위해 사용합니다.
따라서 테스트 대상 코드에서 부수 효과를 일으키는 의존성 시뮬레이션에 가장 유용합니다.

스텁은 함수가 호출되면 미리 정해 놓은 값을 반환하여 함수를 시뮬레이션합니다.  
테스트 대상 코드가 특정 멤버 함수를 호출하여 특정 값을 반환하도록 의존성 시뮬레이션 합니다.
따라서 테스트 대상 코드가 의존하는 코드로 부터 어떤 값을 받아야 하는 경우 유용합니다.

페이크는 클래스의 대체 구현체로, 실제 의존성의 공개 API를 정확하게 시뮬레이션하여 테스트에서 안전하게 사용 가능합니다.
보다 실질적인 테스트가 이루어질 수 있으며 세부 구현 사항 까지 테스트가 잘 확인할 필요가 있을 때 사용합니다.

**Reply by le2sky** [reply]
@seunghye218 

이건 테스트 대역이라기엔 애매한 것 같아요. 마크의 의견은 어떤가용?

**Reply by seunghye218** [reply]
엉뚱한 대답을 했었네요😅 수정했습니다~

**Comment by GIVEN53** [comment]
stub 미리 정의된 값 반환
mock 호출 여부, 순서 검증
fake 대체 구현체, 빠르게 동작
